### PR TITLE
account_activation

### PR DESCRIPTION
**Pull request by weilingooi**
Chapter 11: Account Activation

### CODE REVIEW

**Comment by weilingooi** [comment]
ready

File: app/assets/stylesheets/account_activations.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
```
Comment (chungpv-1008) [inline app/assets/stylesheets/account_activations.scss]: delete this file

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@ def current_user
       @current_user ||= User.find_by id: session[:user_id]    
     elsif user_id = cookies.signed[:user_id]
       user = User.find_by id: user_id
-      if user&.authenticated? cookies[:remember_token]
+      if user&.authenticated? :remember, cookies[:remember_token]
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:11]:
```suggestion
      if user&.authenticated?(:remember, cookies[:remember_token])
```
I'm not sure if you don't have the parenthesis

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation(user)
```
Comment (chungpv-1008) [inline app/mailers/user_mailer.rb]:
```suggestion
  def account_activation user
```

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    @user = user
+
+    mail to: user.email, subject: t(".acc_active")
+  end
+
+  def password_reset
+    @greeting = t"sessions.mail.greeting"
```
Comment (chungpv-1008) [inline app/mailers/user_mailer.rb]:
```suggestion
    @greeting = t "sessions.mail.greeting"
```

File: app/models/user.rb [hunk 5]
```diff
@@ -36,11 +38,34 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    BCrypt::Password.new(remember_digest).is_password? remember_token
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    return false if digest.blank?
```
please research the difference of 'blank' between 'nil'

File: app/models/user.rb [hunk 6]
```diff
@@ -36,11 +38,34 @@ def remember
...
+    BCrypt::Password.new(digest).is_password? token
   end
 
   def forget
     update_attribute :remember_digest, nil
   end
+
+  def activate
+    update_columns(activated: true, activated_at: Time.zone.now)
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    update_columns activated: true, activated_at: Time.zone.now
```

File: app/models/user.rb [hunk 7]
```diff
@@ -36,11 +38,34 @@ def remember
...
+
+  private
+
+  def downcase_email
+    self.email.downcase!
+  end
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest(activation_token)
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    self.activation_digest = User.digest activation_token
```

File: config/routes.rb [hunk 8]
```diff
@@ -10,5 +10,6 @@
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy"
     resources :users, except: %i(new create)
+    resources :account_activations, only: [:edit]
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
    resources :account_activations, only: :edit
```

**Comment by weilingooi** [comment]
ready

File: activation_digest:string [hunk 9]
```diff
@@ -0,0 +1,2 @@
+      invoke  active_record
```
Comment (ngocvt-0484) [inline activation_digest:string]: why use add this file ? 
are u using it?

File: app/models/user.rb [hunk 10]
```diff
@@ -36,11 +38,34 @@ def remember
...
   end
+
+  def activate
+    update_columns activated: true, activated_at: Time.zone.now
+  end
+
+  def send_mail_activate
+    UserMailer.account_activation(self).deliver_now
+  end
+
```
Comment (ngocvt-0484) [inline app/models/user.rb:58]: remove empty line

File: app/models/user.rb [hunk 11]
```diff
@@ -36,11 +38,34 @@ def remember
...
+  def downcase_email
+    self.email.downcase!
+  end
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest activation_token
+  end
 end
+
```
Comment (ngocvt-0484) [inline app/models/user.rb]: remove empty line
just only add end line

**Comment by weilingooi** [comment]
ready

File: app/helpers/account_activations_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/account_activations_helper.rb]: delete file that never used

File: app/mailers/user_mailer.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation user
+    @user = user
+
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove emtpty line

File: app/views/user_mailer/password_reset.html.erb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<h1>User#password_reset</h1>
+
+<p>
+  <%= @greeting %>, find me in app/views/user_mailer/password_reset.html.erb
+</p>
```
Comment (saobv-2223) [inline app/views/user_mailer/password_reset.html.erb]: we don't use reset password in this chapter, therefore delete this file

**Comment by weilingooi** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -1,15 +1,19 @@
...
     
   def create
     user = User.find_by email: params[:session][:email].downcase
     if @user&.authenticate params[:session][:password]
-      flash[:success] = t ".login_sucess"
-      params[:session][:remember] == Settings.remember_me ? remember(@user) : forget(@user)
-      log_in @user
-      redirect_back_or user_path @user
+      if user.activated?
+        params[:session][:remember] == Settings.remember_me ? remember(user) : forget(user)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]:
```suggestion
        params[:session][:remember].eql? Settings.remember_me ? remember(user) : forget(user)
```

File: app/helpers/sessions_helper.rb [hunk 16]
```diff
@@ -8,7 +8,7 @@ def current_user
       @current_user ||= User.find_by id: session[:user_id]    
     elsif user_id = cookies.signed[:user_id]
       user = User.find_by id: user_id
-      if user&.authenticated? cookies[:remember_token]
+      if user&.authenticated?(:remember, cookies[:remember_token])
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: config/environments/production.rb [hunk 17]
```diff
@@ -1,112 +1,40 @@
...
-  # By default Rails will store a last write timestamp in the session. The
-  # DatabaseSelector middleware is designed as such you can define your own
-  # strategy for connection switching and pass that into the middleware through
-  # these configuration options.
-  # config.active_record.database_selector = { delay: 2.seconds }
-  # config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
-  # config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
 end
+
+
```
Comment (saobv-2223) [inline config/environments/production.rb]: add endline

File: config/locales/en.yml [hunk 18]
```diff
@@ -80,11 +80,29 @@ en:
...
+  user_mailer: 
+    acc_active: "Account Activation"
+    hello: "Hello"
+    introduce: "Welcome to Sample App! Click the link below to activate your account:"
+    activate_link: "Activate"
+    acc_heading: "Sample App"
+
+    account_activations:
+    active_fail: "Invalid activation link."
+    active_success: "Account activated!"
```
Comment (saobv-2223) [inline config/locales/en.yml]: add endline

**Comment by weilingooi** [comment]
ready

File: config/locales/en.yml [hunk 19]
```diff
@@ -80,11 +80,30 @@ en:
...
+
```
Comment (manhnt-0960) [inline config/locales/en.yml]: add new line at end of file

File: app/models/user.rb [hunk 20]
```diff
@@ -36,11 +38,32 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    BCrypt::Password.new(remember_digest).is_password? remember_token
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.blank?
```
Comment (manhnt-0960) [inline app/models/user.rb]:
```suggestion
    return false unless digest
```

**Review by manhnt-0960 — COMMENTED** [review]
please fix then merge

**Pull request merged**